src/object/cli_ec.c

```diff
@@ -0,0 +1,443 @@
+					    params->niod.iod_nr);
+		if (rc != 0)
+			return rc;
+		/* Parity is prepended to the recx array, so we have to add
+		 * them here for each encoded stripe.
+		 */
+		D_REALLOC_ARRAY(nrecx, (params->niod.iod_recxs),
+				(params->niod.iod_nr+p));
+		if (nrecx == NULL)
+			return -DER_NOMEM;
+		else if (nrecx != params->niod.iod_recxs)
+			params->niod.iod_recxs = nrecx;
+		for (i = 0; i < p; i++) {
+			params->niod.iod_recxs[params->niod.iod_nr].rx_idx =
+			PARITY_INDICATOR | (s_cur+i*len)/params->niod.iod_size;
+			params->niod.iod_recxs[params->niod.iod_nr++].rx_nr =
+				len / params->niod.iod_size;
+		}
+	}
+	return rc;
+}
+
+/* Updates the params instance for a IOD -- SGL pair.
+ * The parity recxs have already been added, this function appends the
+ * original recx entries.
+ * The parity cells are placed first in the SGL, followed by the
+ * input entries.
+ */
+static int
+ec_update_params(struct ec_params *params, daos_iod_t *iod, d_sg_list_t *sgl,
+		 unsigned int len)
+{
+	daos_recx_t	*nrecx;
+	unsigned int	 i;
+	int		 rc = 0;
+
+	D_REALLOC_ARRAY(nrecx, (params->niod.iod_recxs),
+			(params->niod.iod_nr + iod->iod_nr));
+	if (nrecx == NULL)
+		return -DER_NOMEM;
+	else if (nrecx != params->niod.iod_recxs)
+		params->niod.iod_recxs = nrecx;
+	for (i = 0; i < iod->iod_nr; i++)
+		params->niod.iod_recxs[params->niod.iod_nr++] =
+			iod->iod_recxs[i];
+
+	D_ALLOC_ARRAY(params->nsgl.sg_iovs, (params->p_segs.p_nr + sgl->sg_nr));
+	if (params->nsgl.sg_iovs == NULL)
+		return -DER_NOMEM;
+	for (i = 0; i < params->p_segs.p_nr; i++) {
+		params->nsgl.sg_iovs[i].iov_buf = params->p_segs.p_bufs[i];
+		params->nsgl.sg_iovs[i].iov_buf_len = len;
+		params->nsgl.sg_iovs[i].iov_len = len;
+		params->nsgl.sg_nr++;
+	}
+	for (i = 0; i < sgl->sg_nr; i++)
+		params->nsgl.sg_iovs[params->nsgl.sg_nr++] = sgl->sg_iovs[i];
+
+	return rc;
+}
+
+/* Recover EC allocated memory */
+static void
+ec_free_params(struct ec_params *head)
+{
+	D_FREE(head->iods);
+	D_FREE(head->sgls);
+	while (head != NULL) {
+		int i;
+		struct ec_params *current = head;
+
+		D_FREE(current->niod.iod_recxs);
+		D_FREE(current->nsgl.sg_iovs);
+		for (i = 0; i < current->p_segs.p_nr; i++)
+			D_FREE(current->p_segs.p_bufs[i]);
+		D_FREE(current->p_segs.p_bufs);
+		head = current->next;
+		D_FREE(current);
+	}
+}
+
+
+/* Call-back that recovers EC allocated memory  */
+static int
+ec_free_params_cb(tse_task_t *task, void *data)
+{
+	struct ec_params *head = *((struct ec_params **)data);
+	int		  rc = task->dt_result;
+
+	ec_free_params(head);
+	return rc;
+}
+
+/* Identifies the applicable subset of forwarding targets for non-full-stripe
+ * EC updates.
+ */
+void
+ec_init_tgt_set(daos_iod_t *iods, unsigned int nr,
+		struct daos_oclass_attr *oca, unsigned long *tgt_set)
+{
+	unsigned int    len = oca->u.ec.e_len;
+	unsigned int    k = oca->u.ec.e_k;
+	unsigned int    p = oca->u.ec.e_p;
+	unsigned long	par_only = (1UL << p) - 1;
+	unsigned long	full = (1UL <<  (k+p)) - 1;
```

Comment:
why some place use uint64_t and some other place use unsigned long? all place with uint64_t would be better.

And, could we assume (k+p) <= 64?  probably cannot.

Reply:
It was established in another context that k <= 64, although that context may no longer be valid.  I think it's safe to assume k+p <= 64, but I am willing to discuss it.  (I've not seen a proposal for k > 16 and/or p > 3.

Reply:
> It was established in another context that k <= 64, although that context may no longer be valid. I think it's safe to assume k+p <= 64, but I am willing to discuss it. (I've not seen a proposal for k > 16 and/or p > 3.

GF(2^8) allows k/p's max value as 255.

Personally i think probably unnecessary to add (k+p <= 64) restriction if only for this tgt_set bitmap usage. May refer rebuild's rgt_scan_bits/rgt_pull_bits usage.

On the other hand, I think it not bad to add restriction of (k + p) <= 64 if the real use cases don't need more, But I think it is better to explain it at API header file or some readme file, and check it within obj_ec_codec_init() and return error & err msg for larger EC object class. Without any checking/warning and with internal restriction that will lead to silent failure seems not good.